tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/predicates.md

```diff
@@ -0,0 +1,352 @@
+
+In the `visit_comparison_target` function, we only handle the `column` case where the `path` is empty. The other cases
+will be added when we support the `relationships` capability.
+
+In the `visit_comparison_value` function, we only handle the `scalar` case, in which we push the value onto our
+parameter list. Again, the other cases correspond to capabilities we haven't implemented yet.
+
+The other two expression types are unsupported for now, so we'll throw an error here. We can come back to these later.
+
+```typescript
+// ...
+case "binary_array_comparison_operator":
+  throw new NotSupported("binary_array_comparison_operator is not supported");
+case "exists":
+  throw new NotSupported("exists is not supported");
+// ...
+```
+
+Here's the finished `visit_expression` function:
+
+```typescript
+function visit_expression(parameters: any[], expr: Expression): string {
+    switch (expr.type) {
+        case "and":
+            if (expr.expressions.length > 0) {
+                return expr.expressions.map(e => visit_expression_with_parens(parameters, e)).join(" AND ");
+            } else {
+                return "TRUE";
+            }
+        case "or":
+            if (expr.expressions.length > 0) {
+                return expr.expressions.map(e => visit_expression_with_parens(parameters, e)).join(" OR ");
+            } else {
+                return "FALSE";
+            }
+        case "not":
+            return `NOT ${visit_expression_with_parens(parameters, expr.expression)}`;
+        case "unary_comparison_operator":
+            switch (expr.operator) {
+                case 'is_null':
+                    return `${visit_comparison_target(expr.column)} IS NULL`;
+                default:
+                    throw new BadRequest("Unknown comparison operator");
+            }
+        case "binary_comparison_operator":
+            switch (expr.operator.type) {
+                case 'equal':
+                    return `${visit_comparison_target(expr.column)} = ${visit_comparison_value(parameters, expr.value)}`
+                default:
+                    throw new BadRequest("Unknown comparison operator");
+            }
+        case "binary_array_comparison_operator":
+            throw new NotSupported("binary_array_comparison_operator is not supported");
+        case "exists":
+            throw new NotSupported("exists is not supported");
+        default:
+            throw new BadRequest("Unknown expression type");
+    }
+}
+```
+
+Now let's remove our old snapshots and re-run the test suite.
+
+We can see that predicate tests are passing, but some other test cases are not. That's okay - we'll keep iterating over
+the next few videos until we have all green tests here.
+
+In our snapshots directory, we can also see that we're returning the correct data for some simple predicate queries.
+This query searches for albums for artist with ID `5`, and we can see that the response contains the correct rows.
```

Comment:
It would be helpful to pull out an example response that we expect to see from the `snapshots` directory.

Reply:
Done